### PR TITLE
Ad 3212 adds error handling functionality

### PR DESCRIPTION
**Pull request by thatguynamedandy**
I've reworked this a bit under the following assumptions..

1. Anytime we attempt to get a token and there isn't one, we first attempt to get a token for their backup domain and redirect appropriately, otherwise it's back to login page.

2. When they optionally logout, they always go to the login page, the `store.removeToken()` logs them out of all their apps.

3. There is some new functionality provided by `brandwatchAuthHandleError`, if apps get into an unknown state where there is a valid token but api calls are failing (e.g. 403 / 401 from an api), then they can call this method to follow the steps in point 1.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": ["eslint-config-axiom"],
```
Comment (HHogg) [inline .eslintrc]: While you're here, might as well use the new package `@brandwatch/eslint-config-axiom`.

File: .eslintrc [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  "globals": {
+    "Promise": true,
+    "__BW_REACT_AUTH_DOMAIN__": true,
```
Comment (HHogg) [inline .eslintrc]: Could this be removed, and use `EnvironmentPlugin` in webpack instead?

File: src/BrandwatchReactAuth.js [hunk 3]
```diff
@@ -34,18 +36,28 @@ export default class BrandwatchReactAuth extends Component {
...
+  handleError() {
+    const { backupDomain, backupRedirect } = this.props;
+    const { loginUrl } = this.store;
+
+    if (backupDomain && backupRedirect) {
+      return this.handleGetToken(backupDomain)
+        .then(token => window.location.replace(token ? backupRedirect : loginUrl));
+    }
+
+    return window.location.replace(this.store.loginUrl);
```
Comment (HHogg) [inline src/BrandwatchReactAuth.js]: Is this flow the same as what's happening inside the mount?

File: src/BrandwatchReactAuth.js [hunk 4]
```diff
@@ -1,20 +1,22 @@
...
 import jwtDecode from 'jwt-decode';
 import TokenStore from 'donny-auth';
 
 export default class BrandwatchReactAuth extends Component {
   getChildContext() {
     return {
-      brandwatchAuthLogout: this.handleLogout,
       brandwatchAuthGetProfile: this.handleGetProfile,
       brandwatchAuthGetToken: this.handleGetToken,
+      brandwatchAuthHandleError: this.handleError,
```
Comment (HHogg) [inline src/BrandwatchReactAuth.js]: Now seeing what this does, I'm not sure about the name. It's more of a `handleLogout` but with the `backupRedirect` too?

File: src/BrandwatchReactAuth.js [hunk 5]
```diff
@@ -34,18 +36,28 @@ export default class BrandwatchReactAuth extends Component {
...
     );
   }
 
+  handleError() {
+    const { backupDomain, backupRedirect } = this.props;
+    const { loginUrl } = this.store;
+
+    if (backupDomain && backupRedirect) {
+      return this.handleGetToken(backupDomain)
+        .then(token => window.location.replace(token ? backupRedirect : loginUrl));
```
Comment (HHogg) [inline src/BrandwatchReactAuth.js]: Should we be calling `this.store.removeToken` too?

File: demo/DemoContent.js [hunk 6]
```diff
@@ -37,6 +37,7 @@ export class DemoContent extends Component {
         }
 
         <button id="logout" onClick={ () => brandwatchAuthLogout() }>Logout</button>
+        <button id="error" onClick={ () => brandwatchAuthHandleError() }>Error</button>
```
Comment (pmsorhaindo) [inline demo/DemoContent.js]: Shouldn't we use data-ut-ids here?

File: src/BrandwatchReactAuth.js (hunk 4 above)
Comment (thatguynamedandy) [reply src/BrandwatchReactAuth.js]: Yes good point, logout is a bit odd at the moment, as it only logs you out of the app you are in, really it should remove all tokens. I would be surprised to click logout and be redirected to `my.brandwatch`, as I would then want to log out of that too.

File: src/BrandwatchReactAuth.js (hunk 3 above)
Comment (thatguynamedandy) [reply src/BrandwatchReactAuth.js]: almost, only this gets called regardless of the state of your token, the `mount` flow could be refactored to try and isolate the commonality.

File: demo/DemoContent.js (hunk 6 above)
Comment (thatguynamedandy) [reply demo/DemoContent.js]: yes and maybe no, this is just a demo component and so not used externally. I'm not sure benefits outweigh the extra verbosity needed for the selectors.

**Comment by allouis** [comment]
Looks good to me donny

File: src/BrandwatchReactAuth.js [hunk 7]
```diff
@@ -69,14 +76,16 @@ export default class BrandwatchReactAuth extends Component {
...
   backupRedirect: PropTypes.string,
+  children: PropTypes.node,
+  domain: PropTypes.string.isRequired,
   onCreateStore: PropTypes.func,
 };
 
 BrandwatchReactAuth.childContextTypes = {
   brandwatchAuthGetProfile: PropTypes.func.isRequired,
   brandwatchAuthGetToken: PropTypes.func.isRequired,
+  brandwatchAuthHandleError: PropTypes.func.isRequired,
```
Comment (HHogg) [inline src/BrandwatchReactAuth.js]: I still wonder if this name still describes what it's doing 🤔 It's more of `brandwatchAuthRedirect`?

Comment (thatguynamedandy) [reply src/BrandwatchReactAuth.js]: I see what you mean, or maybe even `brandwatchAuthExitApplication`

**Pull request merged**